### DISCUSSION TITLE
2023-06-02 - Containerized SQL - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by lukecookssw** [SSWConsulting/SSW.Rules.Content#5654] [Chewing the Fat]
Hey team,

Let's talk about [this rule](https://www.ssw.com.au/rules/containerize-sql-server/)

**Comment by duanxinhuan** [comment]
Glad to be the first one.
I am already very confident with Docker.
I might join docker in-depth class for pros

**Comment by leehawkinsssw** [comment]
Containers make this stuff so easy. It's all a far cry from when I had to install different versions of software on different operating systems on removable hard drives!

**Comment by AlexRothwell** [comment]
Managing databases is often quite an involved process so having an easy way to spin up and down databases at your convenience is great. I think this should be the default way of doing local development with a database

**Comment by ozdemir-mehmet** [comment]
Need a PowerPlatform/Dataverse Docker image

**Reply by Hona** [reply]
d365 in a docker image would be next level

**Comment by jakebayliss** [comment]
Don't have much exp with containers so would love to use them more

**Comment by 2fernandez** [comment]
![docker-you-were](https://github.com/SSWConsulting/SSW.Rules.Content/assets/27670973/e3c11888-dc66-4ce0-b6e0-1d0a4abbbecd)

**Comment by Levijj22** [comment]
Docker looks cool.
I thought the video was great at explaining how to use it.

**Comment by UlyssesMaclaren** [comment]
the rule was well written... even I understood it :)

**Comment by leahy268** [comment]
Docker is a great way to be able to run multiple different projects on the same hardware.  Swapping between projects with local machine software must have been a huge pain.

**Comment by tkapa** [comment]
I think that Docker is a valuable to learn, and I have a long way to go with it.

**Comment by danielmackay** [comment]
I like the idea of SQL Containers, however I find local SQL Server slightly easier to manage and use.  This may just be a familiarity thing.

SQL Containers for Integration tests are awesome though!  This allows you to use throwaway DB's that can be used locally and in your build pipelines.

Recently, I've been using the SQL Server Object Explorer within VS to view table schema's and query data.  I find this to be WAY faster than waiting for SSMS to spin up.

![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/2636640/ed4766f7-beeb-4041-b108-62261c85b686)

**Comment by bradystroud** [comment]
I've always used Docker for sql servers 🐳 (Because mac🍎, don't have a choice), so, idk how it works on Windows 🤷‍♂️.
If I switched to windows I'd stick with Docker, because I know how it works.

**Reply by bahjatalaadelssw** [reply]
Go 🍎 Mac !!

**Comment by KahaMason** [comment]
Docker is a great skillset to have. Its cross platform compatibility when dealing with SQL server has come in  handy many times.

**Comment by vladislav-kir** [comment]
Docker is super useful especially for cross-platform dev. I need a lot more practice with it, though, especially with SQL containers, that I never used so far.

**Comment by AshAnil** [comment]
Never used Docker before. The rule shows the benefits of using Docker and it is impressive ⭐ It is very useful for cross-platform devs

**Comment by drwharris** [comment]
Love Containers. SQL in containers is awesome for development.

**Comment by sethdaily** [comment]
Don't know much about this as I'm not a developer - but I can understand why the idea is useful.

**Comment by AntPolkanov** [comment]
I didn't use Docker much, keen to learn more. When is the next Containerization Superpowers workshop? 🙂

**Comment by NadeekaJK** [comment]
Have played with it a bit and like it. Would love a docker super powers for pros.

**Comment by Harry-Ross** [comment]
I've used Docker in the past, and I think its great! It solves a lot of issues not only with databases, but also is great for when projects may target a specific version of a framework.

**Comment by MarcusIrmscherSSW** [comment]
Sounds useful, even with my limited understanding.

**Comment by PennyWalker** [comment]
It was great to learn a little bit more about Dockers and how it works, I think it'll be useful for those who need to use it.

**Comment by jackreimers** [comment]
I've used Docker before for SQL server and prefer that over running a local instance, but I'd like to learn more about using it for things other than SQL.

**Comment by cvionnet** [comment]
Docker concept is really nice, especially to avoid parasitizing your local machine with applications that start tons of Services at each reboot. I didn't know about the 'Azure-Sql-Edge' container.
Love to have more knowledge in this field🎓

**Comment by chrisschultzssw** [comment]
I can see Docker (or alternatives) being super useful. I do have SQL Server and SSMS on my laptop; maybe next time I'll try using a container and Azure Data Studio instead...

**Comment by jeoffreyfischer** [comment]
I've been using them and can see the use for them, but I've come across several instances where they slowed down my work (taking several reboots to connect to container, randomly decided not to connect to VSC anymore, etc.). 
But that just might be me being a noob \o/

**Comment by Marxoz** [comment]
Love containers. Also good for saving resources on your machine too. You can just switch it off when you are not using it, or push it to the cloud when ready. 

I was forced to use it because I use a Mac, but now using it on Windows too.

**Comment by ncn-ssw** [comment]
I have found containers especially useful for experimenting with new software.

**Comment by amankumarrr** [comment]
I recently started using containers and I already love it. It makes it easier to run multiple applications without being impacted by performance issues. Additionally, as mentioned in the video, it's quite easy to spin on and off.

**Comment by steven0x51** [comment]
I have been hearing docker a lot, but haven't actually used a docker. I'd love to have a docker superpower for pros

**Comment by bahjatalaadelssw** [comment]
The future of video editing might be moving to cloud based systems. Nothing local??

**Comment by tomek-i** [comment]
I do enjoy the benefits Docker gives you, especially when working with legacy projects.
You just don't need to worry about having 120 different Node.js version on your machine just to be able to build/run each project that you have been involved in.

It also avoids running into those situations where the local dev works on 4 out of 5 machines and now you got to figure out what's different on that 1 machine that breaks the whole project ... 

I am not that confident with setting up containers and would love a good workshop with the ins and outs and some more advances scenarios including deploying to Azure.

**Comment by wicksipedia** [comment]
Having been burnt by the GAC too many times, I loathe having to install software as dependencies for projects.

I have a strong preference for containerizing app dependencies like the database, storage, caching, etc as the config then gets stored in the repo.

**W**rapped within Docker's platform, every software application enjoys an environment of its own.
**O**vercoming dependency clashes, Docker isolates each application to ensure smooth operations.
**R**eplication of the environment across different systems is seamless, thanks to Docker's containerization.
**K**indly note that its brilliance lies in enabling local development that's reflective of production.
**S**oftware packages and libraries, even specific versions, can be bundled together without hassle.

**O**n top of that, Docker also provides a consistent way to build, package, and deploy applications.
**N**ever again will you hear the phrase, "But it works on my machine." Docker makes this a reality.

**M**irroring production environments locally becomes effortless, reducing development-time bugs.
**Y**ou'll find that it also greatly improves collaboration within teams, by ensuring environment parity.

**M**anaging multiple services running on different software stacks is made possible with Docker Compose.
**A**pplications can be run in isolation, while still being able to communicate with each other.
**C**ontainers are lightweight and start quickly, making Docker an excellent choice for local development.
**H**ence, it brings about an end to the struggle of maintaining different environments for different applications.
**I**n fact, this process of containerization simplifies debugging and testing, facilitating quicker releases.
**N**otably, with Docker, developers can focus on writing code, rather than managing environments.
**E**mbarking on a Docker journey means bringing efficiency, scalability, and reliability to your development workflow.

**Reply by tomek-i** [reply]
WORKS ON MY MACHINE (too) 

I remember those GAC times omg ....


+1 to anything that can be stored in the repo as code/config and then just spun up rather than going through setups and installations

**Comment by lukecookssw** [comment]
Containers for containers' sake is silly. If I need to run something that can't be run natively, then fair enough... But give me local SQL/app instances whenever possible.

**Comment by JackDevAU** [comment]
Docker is great! It can sometimes be a pain to setup but once its setup it creates an easier workflow!

**Comment by Hona** [comment]
I've been hosting in production on a VPS using Docker & Docker compose for many years now.

Only migrated to Azure recently, so I'm learning about ACA & ACS. 

Docker is a great experience & a fun thing to learn. Azure UI for ACA so far looks pretty good, keen to learn more about containers in Azure :)

Happy to help anyone out re: docker & docker compose, especially if you're trying to self host on a VPS

**Comment by tino-liu** [comment]
I think having a dev container really helps with huge project setups. But i've had times where docker just disconnects on me and just won't reconnect resulting in me banging my head trying to get it to work again. Went back to local development right after. Can't say I prefer local over docker at the moment, will explore further once I go into SQL stuff with Docker.

**Comment by Dhruv-0987** [comment]
Its not as easy as it looks the goal is to minimize doing the configuration and get on with development but spinning up the docker itself is a not so easy task and creates a whole different set of issues, sure if once its done then its easy to use but if we can use local infra then why not?

**Comment by zacharykeeping** [comment]
Containers are pretty great for development. I haven't used it for SQL but I think it makes a lot of sense.

**Comment by AttackOnMorty** [comment]
Sometimes people adopt new technologies using the same old ways. They treat containers like virtual machines (Bash into the container and install things manually).

<img width="741" alt="image" src="https://github.com/SSWConsulting/SSW.Rules.Content/assets/37203901/1624034d-2f3c-4147-9ccb-5bb3e1487412">

**Comment by jernejk** [comment]
Docker SQL Server is great but it loses one of the biggest reasons why most projects on my computer work with "Do F5".

Connection string: "data source=.;initial catalog=DB;integrated security=True;"

It just works and I can publish this to GitHub with no worries at all. 😊

With Dockerised SQL Server, you need to manage your credentials, no project with DB is just "Press F5", you need to configure it and I generally don't run a Docker environment until I need it. And haven't needed it for a while.

**Reply by Hona** [reply]
you can use .env files for docker-compose config. You can have a f5 command be docker compose up, and run with no extra setup

**Comment by camillars** [comment]
![giphy](https://github.com/SSWConsulting/SSW.Rules.Content/assets/53064613/7b4bacdd-6495-4dfb-8268-5da5a94b4fb6)

**Comment by jerryluossw** [comment]
Docker SQL Server is great, at the same time, I also like to use docker MySQL and Redis before.
Then I found that starting docker under windows takes up a lot of resources, so I gave up.

**Comment by christoment** [comment]
Azure Docker on Mac works wonder when we have to work on SQL server 😄  - I'm not sold yet on the dev container idea, but perhaps I need more understanding on how that works 😄

**Comment by william-liebenberg** [comment]
Docker gives you wings :)

**Reply by william-liebenberg** [reply]
the isolation and customised infra requirements between projects is awesome. Especially for projects that is more than just an app and a db.

I regularly need to use combination of SQL, Redis, Cosmos, Azurite, Posgresql, Adminer, and now even Dapr for projects. I 100% prefer using a docker-compose (or even dev-container) to share the config with the rest of the team instead a long set of "setup instructions".

**Reply by Hona** [reply]
`docker compose up` with config in something like `.env` is awesome. Makes it easy to see what config you need when deploying to production

**Comment by SamBWagner** [comment]
Containers are cooooool. Never used them though :(

**Comment by KostaMadorsky** [comment]
If configured correctly makes dev experience much better than going through the process of SQL Server installation on the machine.

**Comment by tombui99** [comment]
Docker is an awesome tool and I've been using it to containerize a NodeJS app when working on SSW CodeAuditor. However, sometimes it might not work on certain environment which is annoying but can always be fixed through different configuration. Regardless, I could not think of a better alternative than Docker when it comes to sharing config when working in a big team.

**Comment by AnkyChe** [comment]
The video shows very detailed ways about how to run SQL Server in docker , it helpful (although I didn't understand most part😅)

**Comment by MattParkerDev** [comment]
🤖 
I've had some experience with Docker, and it has definitely simplified my development process. Managing software versions and operating systems used to be a hassle, but Docker makes it much easier. I've found it particularly useful for managing databases during local development. The idea of joining a Docker in-depth class for pros sounds interesting, as it could help me further enhance my skills. Docker has definitely made running multiple projects on the same hardware more convenient. Overall, it's a valuable tool for streamlining development workflows.

**Comment by SofieHong** [comment]
It would be better if all screen displays can be enlarged by at least 125% 🖥️🔍

**Comment by jimmidier** [comment]
I actually enjoyed this new methodology of containerizing infras and apps as much as possible, but it requires more hardware resources than the traditional way of running everything natively. I would definetely turn to this dev flow if I had a 32GB RAM laptop, rather than the poor 16GB I have at the moment.

**Comment by sylhuang** [comment]
I like using docker, especially for spinning up a new project. It has been saving me a lot of time with preset assignment while in uni. But I've never had a chance to use it in production.

**Comment by KristenHu** [comment]
Never used it in production before, but Docker is really cool, hope that I can try it one day.

**Comment by Aibono1225** [comment]
I used Docker once when I was doing a task for SugarLearning. Back then I just followed the instructions to use it and didn't have a full understanding of Docker. I'm glad that this week's CTF allowed me to explore it more. 
Hope to be able to use it in future projects💖.

**Comment by JaneYangssw** [comment]
I love Docker. It allows me to easily create and manage isolated environments for my projects, which saves me a lot of time and effort😊.

**Comment by johnxu16** [comment]
I mainly use containers with Kubernetes for deploying services on Homelab. Using docker on Windows consumes a significant amount of memory and it does slow down other processes, I would rather use it either on a laptop with enough memory or run docker on Linux

**Comment by Ben0189** [comment]
I have limited experience with Docker, would like to expand my knowledge and gain a better understanding of Docker

**Comment by JeanThirion** [comment]
Not a container guru but this is definitely interesting

**Comment by ozair-ashfaque** [comment]
I have less experience with Docker, but keen to learn about it

**Comment by Freego1783** [comment]
I've never really had the opportunity to use Docker, but it seems interesting! Especially for managing multiple versions at the same time.

**Comment by PothieuG** [comment]
I have been able to test Docker on personal projects, and I must say that I was surprised to see how powerful and fast it was.
This is a topic I would like to delve into and/or have the opportunity to work professionally with this tool.

**Comment by BrookJeynes** [comment]
Containers allow for reproducible build on "any" machine, very useful when in a team with many operating systems

**Comment by TobyChurches** [comment]
I haven't really had much experience with docker, but I would be very interested to learn! looks like good stuff 🙂
